File: uniswift/Uni/Uni/Declaration/DeclarationParser.swift
//
//  Parser.swift
//  Uni
//
//  Created by Michael Vlasov on 28.04.2018.
//  Copyright © 2018 Unified. All rights reserved.
//

import Foundation

public class DeclarationParser {
	
	public static func parse(string: String) throws -> [DeclarationElement] {
		return try parse(chars: CharReader(string: string))
	}
	

	public static func parse(chars: CharReader) throws -> [DeclarationElement] {
		let tokens = try TokenReader(chars: chars, tokenizer: tokenize)
		var elements = [DeclarationElement]()
		if try tokens.pass(token: indentation) && tokens.passed.valueInt == 0 {
			try parse(elements: &elements, indentation: 0, tokens: tokens)
		}
		if tokens.hasCurrent {
			throw ParsingError.unexpected(token: tokens.current, context: tokens.currentContext)
		}
		return elements
	}

	// Internals
	
	private static func isNotCrOrLf(c: UnicodeScalar) -> Bool {
		return c != "\r" && c != "\n"
	}
	
	
	private static func isTab(c: UnicodeScalar) -> Bool {
		return c == "\t"
	}
	
	
	private static let reservedUnicode = " \t\r\n()='`~#".unicodeScalars
	private static func isNotReserved(c: UnicodeScalar) -> Bool {
		return !reservedUnicode.contains(c)
	}
	
	
	private static func isWhitespace(c: UnicodeScalar) -> Bool {
		return c == " " || c == "\t"
	}
	
	
	private static func parseBackQuotedString(chars: CharReader) throws -> LexicalToken {
		var string = ""
		let startPos = chars.currentPos
		while chars.hasCurrent {
			chars.passWhile { $0 != "`" }
			string += chars.passed
			if !chars.passChar("`") {
				throw ParsingError.error(message: "Unterminated backquoted string", context: chars.contextAt(pos: startPos))
			}
			if !chars.passChar("`") {
				break
			}
			string += "`"
		}
		return nameOrValue.with(value: .string(string))
	}
	
	
	private static let escapedByChar: [UnicodeScalar: UnicodeScalar] = [
		"0": "\0",
		"\'": "\'",
		"\\": "\\",
		"n": "\n",
		"r": "\r",
		"t": "\t"]
	
	private static let hexChars = CharacterSet(charactersIn: "0123456789ABCDEFabcdef")
	
	
	private static func expectCharFromCodePoint(chars: CharReader, digits: Int) throws -> UnicodeScalar{
		var hex = [UnicodeScalar]()
		var digits = digits
		while digits > 0 {
			if !chars.hasCurrent || !hexChars.contains(chars.current) {
				throw ParsingError.error(message: "Invalid hex code string", context: chars.contextAt(pos: chars.currentPos))
			}
			hex.append(chars.current)
			chars.next()
			digits -= 1
		}
		let hexString = String(String.UnicodeScalarView(hex))
		return UnicodeScalar(Int(hexString, radix: 16) ?? 0) ?? "\0"
	}
	
	
	private static func parseSingleQuotedString(chars: CharReader) throws -> LexicalToken {
		var unicode = [UnicodeScalar]()
		let startPos = chars.currentPos
		while chars.hasCurrent {
			if chars.passChar("\'") {
				return nameOrValue.with(value: .string(String(String.UnicodeScalarView(unicode))))
			}
			if chars.passChar("\\") {
				if !chars.hasCurrent {
					break
				}
				if let escaped = escapedByChar[chars.current] {
					unicode.append(escaped)
					chars.next()
				} else if chars.passChar("x") {
					unicode.append(try expectCharFromCodePoint(chars: chars, digits: 2))
				} else if chars.passChar("u") {
					unicode.append(try expectCharFromCodePoint(chars: chars, digits: 4))
				} else {
					throw ParsingError.error(message: "Invalid escape sequence", context: chars.currentContext)
				}
			} else {
				unicode.append(chars.current)
				chars.next()
			}
		}
		throw ParsingError.error(message: "Unterminated backquoted string", context: chars.contextAt(pos: startPos))
	}
	
	
	private static func passLineEnd(chars: CharReader) -> Bool {
		if chars.current == "#" {
			chars.passWhile(isNotCrOrLf)
		}
		if chars.passChar("\r") {
			chars.passChar("\n")
			return true
		}
		if chars.passChar("\n") {
			return true
		}
		return false
	}
	
	
	private static func parseIndentation(chars: CharReader) -> LexicalToken? {
		while chars.tag == nil || passLineEnd(chars: chars) {
			chars.tag = true
			let start = chars.nextPos
			chars.passWhile(isTab)
			if !chars.hasCurrent {
				return nil
			}
			chars.passWhile(isWhitespace)
			if chars.current != "\r" && chars.current != "\n" && chars.current != "#" {
				let level = chars.unicode.distance(from: start, to: chars.nextPos)
				return indentation.with(value: .int(level))
			}
		}
		return nil
	}
	
	
	private static func tokenize(chars: CharReader) throws -> LexicalToken? {
		var token = parseIndentation(chars: chars)
		if let token = token {
			return token
		}
		guard chars.hasCurrent else {
			return nil
		}
		if chars.passChar("=") {
			token = nameValueSeparator
		} else if chars.passChar("~") {
			token = attributeContinuation
		} else if chars.passChar("(") {
			token = openList
		} else if chars.passChar(")") {
			token = closeList
		} else if chars.passChar("\'") {
			token = try parseSingleQuotedString(chars: chars)
		} else if chars.passChar("`") {
			token = try parseBackQuotedString(chars: chars)
		} else if chars.passWhile(isNotReserved) {
			token = nameOrValue.with(value: .string(String(chars.passed)))
		} else {
			return nil
		}
		chars.passWhile(isWhitespace)
		return token
	}
	
	
	private static func parse(attributes: inout [DeclarationAttribute], tokens: TokenReader) throws {
		while try tokens.pass(token: nameOrValue) {
			let name = tokens.passed.valueString
			var value = DeclarationAttribute.Value.missing
			if try tokens.pass(token: nameValueSeparator) {
				if try tokens.pass(token: openList) {
					var items = [String]()
					while try tokens.pass(token: nameOrValue) {
						items.append(tokens.passed.valueString)
					}
					try tokens.pass(required: closeList)
					value = .array(items)
				} else {
					try tokens.pass(required: nameOrValue)
					value = .string(tokens.passed.valueString)
				}
			}
			attributes.append(DeclarationAttribute(name: name, value: value))
		}
	}
	
	
	private static func pass(expectedIndentation: Int, tokens: TokenReader) throws -> Bool {
		if tokens.hasCurrent && tokens.current.name == indentation.name &&
			tokens.current.valueInt == expectedIndentation {
			try tokens.next()
			return true
		}
		return false
	}
	
	
	private static func parseElement(to elements: inout [DeclarationElement], indentation: Int, tokens: TokenReader) throws  {
		var attributes = [DeclarationAttribute]()
		var children = [DeclarationElement]()
		try parse(attributes: &attributes, tokens: tokens)
		while try pass(expectedIndentation: indentation + 1, tokens: tokens) {
			if try tokens.pass(token: attributeContinuation) {
				try parse(attributes: &attributes, tokens: tokens)
			} else {
				try parse(elements: &children, indentation: indentation + 1, tokens: tokens)
				break
			}
		}
		elements.append(DeclarationElement(attributes: attributes, children: children))
	}
	
	
	private static func parse(elements: inout [DeclarationElement], indentation: Int, tokens: TokenReader) throws {
		repeat {
			try parseElement(to: &elements, indentation: indentation, tokens: tokens)
		} while try pass(expectedIndentation: indentation, tokens: tokens)
	}
	

	static let nameOrValue = LexicalToken.register("name or value")
	static let nameValueSeparator = LexicalToken.register("=")
	static let attributeContinuation = LexicalToken.register("~")
	static let indentation = LexicalToken.register("⇥")
	static let openList = LexicalToken.register("(")
	static let closeList = LexicalToken.register(")")
}

File: unijava/src/org/unified/uni/declaration/DeclarationParser.java
package org.unified.uni.declaration;

import org.unified.uni.lexer.CharReader;
import org.unified.uni.lexer.LexicalToken;
import org.unified.uni.lexer.TokenReader;

import java.util.ArrayList;
import java.util.HashMap;

/**
 * Uni Declaration Parser
 * <p>
 * Created by Michael Vlasov on 19.04.2018.
 * <p>
 * Copyright (c) 2018 Michael Vlasov
 */

public class DeclarationParser {

	public static DeclarationElement[] parse(String source) throws TokenReader.Error {
		return parse(new CharReader(source));
	}


	public static DeclarationElement[] parse(CharReader chars) throws TokenReader.Error {
		TokenReader tokens = new TokenReader(chars, DeclarationParser::tokenize);
		ArrayList<DeclarationElement> elements = new ArrayList<>();
		if (tokens.pass(indentation) && tokens.getPassed().valueInt() == 0) {
			parseElements(tokens, elements, 0);
		}
		if (tokens.isHasCurrent()) {
			throw new TokenReader.Error("Unexpected \"" + tokens.getCurrent().toString() + "\" at " + tokens.currentContext());
		}
		return elements.toArray(DeclarationElement.emptyArray);
	}


	// Internals


	private static boolean isNotCrOrLf(char c) {
		return c != '\r' && c != '\n';
	}


	private static boolean isTab(char c) {
		return c == '\t';
	}


	private static boolean isNotReserved(char c) {
		return " \t\r\n()='`~#".indexOf(c) < 0;
	}


	private static boolean isWhitespace(char c) {
		return (c == ' ') || (c == '\t');
	}


	private static LexicalToken parseBackquotedString(CharReader chars) throws TokenReader.Error {
		StringBuilder string = new StringBuilder();
		int startPos = chars.getCurrentPos();
		while (chars.isHasCurrent()) {
			chars.passWhile((c) -> c != '`');
			string.append(chars.passed());
			if (!chars.passChar('`')) {
				throw new TokenReader.Error("Unterminated backquoted string at " + chars.contextAt(startPos));
			}
			if (!chars.passChar('`')) {
				break;
			}
			string.append('`');
		}
		return nameOrValue.with(string.toString());
	}


	private static HashMap<Character, Character> escapedByChar = new HashMap<Character, Character>() {{
		put('0', '\0');
		put('\'', '\'');
		put('\\', '\\');
		put('n', '\n');
		put('r', '\r');
		put('t', '\t');
		put('b', '\b');
		put('f', '\f');
	}};


	private static char expectCharFromCodePoint(CharReader chars, int digits) throws TokenReader.Error {
		StringBuilder hex = new StringBuilder();
		while (digits > 0) {
			String hexChars = "0123456789ABCDEFabcdef";
			if (!chars.isHasCurrent() || hexChars.indexOf(chars.getCurrent()) < 0) {
				throw new TokenReader.Error("Invalid hex code string at " + chars.contextAt(chars.getCurrentPos()));
			}
			hex.append(chars.getCurrent());
			chars.next();
			digits -= 1;
		}
		return Character.toChars(Integer.parseInt(hex.toString(), 16))[0];
	}


	private static LexicalToken parseSinglequotedString(CharReader chars) throws TokenReader.Error {
		StringBuilder string = new StringBuilder();
		int startPos = chars.getCurrentPos();
		while (chars.isHasCurrent()) {
			if (chars.passChar('\'')) {
				return nameOrValue.with(string.toString());
			}
			if (chars.passChar('\\')) {
				if (!chars.isHasCurrent()) {
					break;
				}
				Character escaped = escapedByChar.get(chars.getCurrent());
				if (escaped != null) {
					string.append(escaped);
					chars.next();
				} else if (chars.passChar('x')) {
					string.append(expectCharFromCodePoint(chars, 2));
				} else if (chars.passChar('u')) {
					string.append(expectCharFromCodePoint(chars, 4));
				} else {
					throw new TokenReader.Error("Invalid escape sequence at " + chars.currentContext());
				}
			} else {
				string.append(chars.getCurrent());
				chars.next();
			}
		}
		throw new TokenReader.Error("Unterminated backquoted string at " + chars.contextAt(startPos));
	}


	private static boolean passLineEnd(CharReader chars) {
		if (chars.getCurrent() == '#') {
			chars.passWhile(DeclarationParser::isNotCrOrLf);
		}
		if (chars.passChar('\r')) {
			chars.passChar('\n');
			return true;
		}
		return chars.passChar('\n');
	}


	private static LexicalToken parseIndentation(CharReader chars) {
		while (chars.getTag() == null || passLineEnd(chars)) {
			chars.setTag(true);
			int start = chars.getNextPos();
			chars.passWhile(DeclarationParser::isTab);
			if (!chars.isHasCurrent()) {
				return null;
			}
			chars.passWhile(DeclarationParser::isWhitespace);
			char c = chars.getCurrent();
			if (c != '\r' && c != '\n' && c != '#') {
				return indentation.with(chars.getNextPos() - start);
			}
		}
		return null;
	}


	private static LexicalToken tokenize(CharReader chars) throws TokenReader.Error {
		LexicalToken token = parseIndentation(chars);
		if (token != null) {
			return token;
		}
		if (!chars.isHasCurrent()) {
			return null;
		}
		if (chars.passChar('=')) {
			token = nameValueSeparator;
		} else if (chars.passChar('~')) {
			token = attributeContinuation;
		} else if (chars.passChar('(')) {
			token = openList;
		} else if (chars.passChar(')')) {
			token = closeList;
		} else if (chars.passChar('\'')) {
			token = parseSinglequotedString(chars);
		} else if (chars.passChar('`')) {
			token = parseBackquotedString(chars);
		} else if (chars.passWhile(DeclarationParser::isNotReserved)) {
			token = nameOrValue.with(chars.passed());
		} else {
			return null;
		}
		chars.passWhile(DeclarationParser::isWhitespace);
		return token;
	}


	private static void parseAttributes(TokenReader tokens, ArrayList<DeclarationAttribute> attributes) throws TokenReader.Error {
		while (tokens.pass(nameOrValue)) {
			String name = tokens.getPassed().valueString();
			Object value = null;
			if (tokens.pass(nameValueSeparator)) {
				if (tokens.pass(openList)) {
					ArrayList<String> arrayBuilder = new ArrayList<>();
					while (tokens.pass(nameOrValue)) {
						arrayBuilder.add(tokens.getPassed().valueString());
					}
					tokens.passRequired(closeList);
					value = arrayBuilder.toArray();
				} else {
					tokens.passRequired(nameOrValue);
					value = tokens.getPassed().valueString();
				}
			}
			attributes.add(new DeclarationAttribute(name, value));
		}
	}


	private static boolean passIndentation(TokenReader tokens, int expectedIndentation) throws TokenReader.Error {
		if (tokens.isHasCurrent() && tokens.getCurrent().name == indentation.name &&
			tokens.getCurrent().valueInt() == expectedIndentation) {
			tokens.next();
			return true;
		}
		return false;
	}


	private static void parseElement(TokenReader tokens, ArrayList<DeclarationElement> elements, int indentation) throws TokenReader.Error {
		ArrayList<DeclarationAttribute> attributes = new ArrayList<>();
		ArrayList<DeclarationElement> children = new ArrayList<>();
		parseAttributes(tokens, attributes);
		while (passIndentation(tokens, indentation + 1)) {
			if (tokens.pass(attributeContinuation)) {
				parseAttributes(tokens, attributes);
			} else {
				parseElements(tokens, children, indentation + 1);
				break;
			}
		}
		elements.add(new DeclarationElement(attributes.toArray(DeclarationAttribute.emptyArray), children.toArray(DeclarationElement.emptyArray)));
	}


	private static void parseElements(TokenReader tokens, ArrayList<DeclarationElement> elements, int indentation) throws TokenReader.Error {
		do {
			parseElement(tokens, elements, indentation);
		} while (passIndentation(tokens, indentation));
	}


	private static LexicalToken nameOrValue = LexicalToken.register("name or value");
	private static LexicalToken nameValueSeparator = LexicalToken.register("=");
	private static LexicalToken attributeContinuation = LexicalToken.register("~");
	private static LexicalToken indentation = LexicalToken.register("⇥");
	private static LexicalToken openList = LexicalToken.register("(");
	private static LexicalToken closeList = LexicalToken.register(")");

}
